File: yams/cpp/src/yams/motorcontrollers/simulation/ElevatorSimSupplier.cpp
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

#include "yams/motorcontrollers/simulation/ElevatorSimSupplier.hpp"

#include <wpi/simulation/RoboRioSim.hpp>

#include <utility>

namespace yams::motorcontrollers::simulation {

ElevatorSimSupplier::ElevatorSimSupplier(wpi::sim::ElevatorSim& sim,
                                         std::function<double()> dutyCycleSupplier,
                                         const gearing::MechanismGearing& gearing,
                                         wpi::units::meter_t circumference, wpi::units::second_t period)
    : m_sim(sim),
      m_dutyCycleSupplier(std::move(dutyCycleSupplier)),
      m_gearing(gearing),
      m_circumference(circumference),
      m_period(period) {}

void ElevatorSimSupplier::UpdateSim() {
  if (!m_inputFed) {
    m_lastInputVoltage = wpi::units::volt_t{m_dutyCycleSupplier() * GetMechanismSupplyVoltage().value()};
    m_sim.SetInputVoltage(m_lastInputVoltage);
  }
  m_inputFed = false;
  m_sim.Update(m_period);
  StarveWatchdog();
}

wpi::units::turn_t ElevatorSimSupplier::GetMechanismPosition() {
  return LinearToMechanismAngle(m_sim.GetPosition());
}

wpi::units::turns_per_second_t ElevatorSimSupplier::GetMechanismVelocity() {
  return LinearToMechanismVelocity(m_sim.GetVelocity());
}

wpi::units::turns_per_second_squared_t ElevatorSimSupplier::GetMechanismAcceleration() {
  return wpi::units::turns_per_second_squared_t{0.0};
}

wpi::units::turn_t ElevatorSimSupplier::GetRotorPosition() {
  return GetMechanismPosition() * m_gearing.GetMechanismToRotorRatio();
}

wpi::units::turns_per_second_t ElevatorSimSupplier::GetRotorVelocity() {
  return GetMechanismVelocity() * m_gearing.GetMechanismToRotorRatio();
}

wpi::units::turns_per_second_squared_t ElevatorSimSupplier::GetRotorAcceleration() {
  return wpi::units::turns_per_second_squared_t{0.0};
}

void ElevatorSimSupplier::SetMechanismPosition(wpi::units::turn_t angle) {
  m_sim.SetState(MechanismAngleToLinear(angle), m_sim.GetVelocity());
}

void ElevatorSimSupplier::SetMechanismVelocity(wpi::units::turns_per_second_t velocity) {
  m_sim.SetState(m_sim.GetPosition(), MechanismVelocityToLinear(velocity));
}

void ElevatorSimSupplier::SetRotorPosition(wpi::units::turn_t angle) {
  SetMechanismPosition(angle / m_gearing.GetMechanismToRotorRatio());
}

void ElevatorSimSupplier::SetRotorVelocity(wpi::units::turns_per_second_t velocity) {
  SetMechanismVelocity(velocity / m_gearing.GetMechanismToRotorRatio());
}

bool ElevatorSimSupplier::IsWatchdogExpired() { return !m_watchdogFed; }

void ElevatorSimSupplier::FeedWatchdog() { m_watchdogFed = true; }

void ElevatorSimSupplier::StarveWatchdog() { m_watchdogFed = false; }

wpi::units::ampere_t ElevatorSimSupplier::GetCurrentDrawAmps() { return m_sim.GetCurrentDraw(); }

void ElevatorSimSupplier::SetInputVoltage(wpi::units::volt_t volts) {
  m_lastInputVoltage = volts;
  m_sim.SetInputVoltage(volts);
  m_inputFed = true;
  FeedWatchdog();
}

wpi::units::volt_t ElevatorSimSupplier::GetMechanismSupplyVoltage() {
  return wpi::sim::RoboRioSim::GetVInVoltage();
}

wpi::units::volt_t ElevatorSimSupplier::GetMechanismStatorVoltage() { return m_lastInputVoltage; }

void ElevatorSimSupplier::SetMechanismStatorVoltage(wpi::units::volt_t volts) { SetInputVoltage(volts); }

// ---- Private helpers --------------------------------------------------------

wpi::units::turn_t ElevatorSimSupplier::LinearToMechanismAngle(wpi::units::meter_t position) const {
  return wpi::units::turn_t{position.value() / m_circumference.value()};
}

wpi::units::turns_per_second_t ElevatorSimSupplier::LinearToMechanismVelocity(
    wpi::units::meters_per_second_t velocity) const {
  return wpi::units::turns_per_second_t{velocity.value() / m_circumference.value()};
}

wpi::units::meter_t ElevatorSimSupplier::MechanismAngleToLinear(wpi::units::turn_t angle) const {
  return wpi::units::meter_t{angle.value() * m_circumference.value()};
}

wpi::units::meters_per_second_t ElevatorSimSupplier::MechanismVelocityToLinear(
    wpi::units::turns_per_second_t velocity) const {
  return wpi::units::meters_per_second_t{velocity.value() * m_circumference.value()};
}

}  // namespace yams::motorcontrollers::simulation

File: yams/java/yams/motorcontrollers/simulation/ElevatorSimSupplier.java
// Copyright (c) 2026 Yet Another Software Suite
// SPDX-License-Identifier: LGPL-3.0-or-later

package yams.motorcontrollers.simulation;

import static org.wpilib.units.Units.Amps;
import static org.wpilib.units.Units.Meters;
import static org.wpilib.units.Units.MetersPerSecond;
import static org.wpilib.units.Units.Microsecond;
import static org.wpilib.units.Units.Milliseconds;
import static org.wpilib.units.Units.RadiansPerSecond;
import static org.wpilib.units.Units.RotationsPerSecond;
import static org.wpilib.units.Units.Seconds;
import static org.wpilib.units.Units.Volts;

import org.wpilib.math.system.DCMotor;
import org.wpilib.units.measure.Angle;
import org.wpilib.units.measure.AngularAcceleration;
import org.wpilib.units.measure.AngularVelocity;
import org.wpilib.units.measure.Current;
import org.wpilib.units.measure.Voltage;
import org.wpilib.simulation.ElevatorSim;
import org.wpilib.simulation.RoboRioSim;
import java.util.UUID;
import java.util.function.Supplier;
import yams.gearing.MechanismGearing;
import yams.math.DerivativeTimeFilter;
import yams.motorcontrollers.SimSupplier;
import yams.motorcontrollers.SmartMotorController;
import yams.motorcontrollers.SmartMotorControllerConfig;

/**
 * ElevatorSim Supplier — simulates an elevator mechanism using WPILib's
 * {@link org.wpilib.simulation.ElevatorSim}.
 *
 * <p>
 * This supplier steps WPILib's {@code ElevatorSim} physics model each control loop and exposes
 * the resulting height, linear velocity, current draw, and voltage through the
 * {@link yams.motorcontrollers.SimSupplier} interface. Because {@code ElevatorSim} operates in
 * linear units (meters), positions and velocities are converted to and from mechanism (angular)
 * units using the associated {@link yams.motorcontrollers.SmartMotorController}'s config.
 * </p>
 *
 * <h2>Example</h2>
 * <pre>{@code
 * // 1. Build the WPILib elevator physics model
 * ElevatorSim elevatorPhysics = new ElevatorSim(
 *     DCMotor.getNEO(1),
 *     10.0,                 // gear ratio (rotor/mechanism)
 *     5.0,                  // carriage mass (kg)
 *     Units.inchesToMeters(1.0), // drum radius (meters)
 *     0.0,                  // minimum height (meters)
 *     1.5,                  // maximum height (meters)
 *     true,                 // simulate gravity
 *     0.0);                 // starting height (meters)
 *
 * // 2. Configure and build the YAMS smart motor controller
 * SmartMotorController motor = new SparkMaxController(
 *     new SmartMotorControllerConfig()
 *         .withGearing(new MechanismGearing(10.0))
 *         .withClosedLoopControlPeriod(Milliseconds.of(20)));
 *
 * // 3. Wrap physics model in the supplier and register it
 * ElevatorSimSupplier sim = new ElevatorSimSupplier(elevatorPhysics, motor);
 * motor.getConfig().withSimSupplier(sim);
 * }</pre>
 */
public class ElevatorSimSupplier implements SimSupplier {
  private final ElevatorSim sim;
  private final SmartMotorControllerConfig config;
  private final MechanismGearing mechGearing;
  private final DCMotor motor;
  private final UUID uuid;
  private final Supplier<Double> motorDutyCycleSupplier;
  private final Supplier<Double> pos;
  private final Supplier<Double> mps;
  private final DerivativeTimeFilter mpsps;
  private boolean inputFed = false;
  private boolean simUpdated = false;

  /**
   * Construct the ElevatorSim supplier
   *
   * @param simulation           Simulation instance
   * @param smartMotorController SMC for the ElevatorSim.
   */
  public ElevatorSimSupplier(ElevatorSim simulation, SmartMotorController smartMotorController) {
    sim = simulation;
    config = smartMotorController.getConfig();
    mechGearing = config.getGearing();
    motor = smartMotorController.getDCMotor();
    uuid = smartMotorController.m_batterySimUUID;
    motorDutyCycleSupplier = smartMotorController::getDutyCycle;
    pos = sim::getPosition;
    mps = sim::getVelocity;
    mpsps = new DerivativeTimeFilter(
        pos.get(), config.getClosedLoopControlPeriod().orElse(Milliseconds.of(20)));
  }

  @Override
  public void updateSimState() {
    if (!isInputFed()) {
      sim.setInputVoltage(motorDutyCycleSupplier.get() * RoboRioSim.getVInVoltage());
      RoboRioSim.setVInVoltage(BatterySim.calculateVoltage(uuid, sim.getCurrentDraw()));
    }
    if (!simUpdated) {
      starveInput();
      sim.update(config.getClosedLoopControlPeriod().orElse(Milliseconds.of(20)).in(Seconds));
      feedUpdateSim();
    }
  }

  @Override
  public boolean getUpdatedSim() {
    return simUpdated;
  }

  @Override
  public void feedUpdateSim() {
    simUpdated = true;
  }

  @Override
  public void starveUpdateSim() {
    simUpdated = false;
  }

  @Override
  public boolean isInputFed() {
    return inputFed;
  }

  @Override
  public void feedInput() {
    inputFed = true;
  }

  @Override
  public void starveInput() {
    inputFed = false;
  }

  @Override
  public void setMechanismStatorDutyCycle(double dutyCycle) {
    feedInput();
    sim.setInputVoltage(dutyCycle * getMechanismSupplyVoltage().in(Volts));
  }

  @Override
  public Voltage getMechanismSupplyVoltage() {
    return Volts.of(RoboRioSim.getVInVoltage());
  }

  @Override
  public Voltage getMechanismStatorVoltage() {
    return Volts.of(motor.getVoltage(
        motor.getTorque(sim.getCurrentDraw()), getMechanismVelocity().in(RadiansPerSecond)));
  }

  @Override
  public void setMechanismStatorVoltage(Voltage volts) {
    feedInput();
    sim.setInputVoltage(volts.in(Volts));
  }

  @Override
  public Angle getMechanismPosition() {
    return config.convertToMechanism(Meters.of(pos.get()));
  }

  @Override
  public void setMechanismPosition(Angle position) {
    sim.setState(config.convertFromMechanism(position).in(Meters), mps.get());
  }

  @Override
  public Angle getRotorPosition() {
    return getMechanismPosition().times(mechGearing.getMechanismToRotorRatio());
  }

  @Override
  public AngularVelocity getMechanismVelocity() {
    return config.convertToMechanism(MetersPerSecond.of(mps.get()));
  }

  @Override
  public void setMechanismVelocity(AngularVelocity velocity) {
    sim.setState(pos.get(), config.convertFromMechanism(velocity).in(MetersPerSecond));
  }

  @Override
  public AngularVelocity getRotorVelocity() {
    return getMechanismVelocity().times(mechGearing.getMechanismToRotorRatio());
  }

  @Override
  public Current getCurrentDraw() {
    return Amps.of(sim.getCurrentDraw());
  }

  @Override
  public AngularAcceleration getRotorAcceleration() {
    return RotationsPerSecond.per(Microsecond)
        .of(mpsps.derivative(getRotorVelocity().in(RotationsPerSecond)));
  }
}
